### PR TITLE
Enforce Certificate Transparency policy for SSL connections

### PR DESCRIPTION
**Pull request by RJPercival**
If a ct_validation_callback is set on SSL or SSL_CTX, SCTs (Signed Certificate Timestamps) will be requested when establishing an SSL connection. Their signatures will be verified and they will be sorted into good and bad SCTs, which are then handed over to the callback. It can then decide whether to continue the connection or abort it, based on the SCTs provided.

s_client has been extended to allow a few basic policies to be enabled:
- noct: SCTs will be neither requested nor verified, even if provided.
- requestct: requires only that no SCTs fail verification (e.g. providing no SCTs is acceptable).
- requirect: requires that at least one good SCT is provided.

I have alternative tests in another branch (https://github.com/RJPercival/openssl/commit/2ec37816d5ff2477725a353102e9f972f6ddd510), which use a larger collection of test data but do less thorough tests. It would be helpful to get opinions on how these should be merged with the existing tests.

### CODE REVIEW

**Comment by benlaurie** [comment]
I needed to do this to make it build/test:
```
diff --git a/crypto/ct/ct_log.c b/crypto/ct/ct_log.c
index 3394547..0189b1c 100644
--- a/crypto/ct/ct_log.c
+++ b/crypto/ct/ct_log.c
@@ -67,8 +67,6 @@ struct ctlog_st {
     EVP_PKEY *public_key;
 };
 
-DEFINE_STACK_OF(CTLOG)
-
 struct ctlog_store_st {
     STACK_OF(CTLOG) *logs;
 };
diff --git a/crypto/ct/ct_x509v3.c b/crypto/ct/ct_x509v3.c
index 011c7c5..9659cfc 100644
--- a/crypto/ct/ct_x509v3.c
+++ b/crypto/ct/ct_x509v3.c
@@ -64,6 +64,7 @@
 #include <openssl/ct.h>
 #include <openssl/obj_mac.h>
 #include <openssl/x509v3.h>
+#include <internal/ct_int.h>
 
 static char *i2s_poison(const X509V3_EXT_METHOD *method, void *val)
 {
diff --git a/include/openssl/ct.h b/include/openssl/ct.h
index 820074a..1164b76 100644
--- a/include/openssl/ct.h
+++ b/include/openssl/ct.h
@@ -111,6 +111,7 @@ typedef struct sct_ctx_st SCT_CTX;
 
 /* CT log information */
 typedef struct ctlog_st CTLOG;
+DEFINE_STACK_OF(CTLOG)
 
 /*
  * A store for multiple CTLOG instances.
```

File: crypto/ct/ct_b64.c [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    outlen = (inlen / 4) * 3;
+    outbuf = OPENSSL_malloc(outlen);
+    if (outbuf == NULL) {
+        CTerr(CT_F_CT_BASE64_DECODE, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    outlen = EVP_DecodeBlock(outbuf, (unsigned char *)in, inlen);
+    if (outlen < 0) {
+        OPENSSL_free(outbuf);
```
Comment (ajmohan) [inline crypto/ct/ct_b64.c]: outbuf is freed second time at err:

File: crypto/ct/ct_b64.c [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    int pkey_der_len;
+    EVP_PKEY *pkey = NULL;
+    CTLOG *log = NULL;
+
+    pkey_der_len = CT_base64_decode(pkey_base64, &pkey_der);
+    if (pkey_der_len <= 0) {
+        CTerr(CT_F_CTLOG_NEW_FROM_BASE64, CT_R_LOG_CONF_INVALID_KEY);
+        return NULL;
+    }
+
```
Comment (ajmohan) [inline crypto/ct/ct_b64.c]: pkey_der need to freed after usage.

Comment (RJPercival) [inline crypto/ct/ct_b64.c]: Done.

File: crypto/ct/ct_b64.c (hunk 1 above)
Comment (RJPercival) [inline crypto/ct/ct_b64.c]: Fixed.

**Comment by RJPercival** [comment]
Ben's fixes were integrated into an earlier PR, so they are no longer necessary. This PR is ready for review.

**Comment by richsalz** [comment]
merging tests won't change the API so that can wait.  need doc updates for the new flags.  need to possibly rebase after @ekasper handshake stuff goes in.

**Comment by richsalz** [comment]
also, do we need other docs on the SCT accessor funcs?

**Comment by RJPercival** [comment]
Rebased and added documentation for the s_client flags.

**Comment by richsalz** [comment]
docs for the set_ct_validation functions?

**Comment by RJPercival** [comment]
Documentation for the new SSL functions added.

**Comment by richsalz** [comment]
Thanks.  Sorry for doing this in drips and drabs, but busy at RSA and want to get this done before the freeze.

**Comment by richsalz** [comment]
+1
Well merge after ... someone else ... reviews.

File: apps/Makefile.in [hunk 3]
```diff
@@ -92,10 +93,14 @@ install:
...
-	chmod 644 $(DESTDIR)$(OPENSSLDIR)/openssl.cnf.new; \
-	mv -f  $(DESTDIR)$(OPENSSLDIR)/openssl.cnf.new $(DESTDIR)$(OPENSSLDIR)/openssl.cnf
+	 done;
+	@set -e; for i in $(CONFS); \
+	do \
+	(echo installing $$i; \
+	 cp $$i $(DESTDIR)$(OPENSSLDIR)/$$i.new; \
+	 chmod 644 $(DESTDIR)$(OPENSSLDIR)/$$i.new; \
+	 mv -f $(DESTDIR)$(OPENSSLDIR)/$$i.new $(DESTDIR)$(OPENSSLDIR)/$$i ); \
+	 done;
```
Comment (benlaurie) [inline apps/Makefile.in]: ; on done redundant?

File: apps/Makefile.in [hunk 4]
```diff
@@ -107,8 +112,12 @@ uninstall:
...
 		echo $(RM) $(DESTDIR)$(OPENSSLDIR)/misc/$$i; \
 		$(RM) $(DESTDIR)$(OPENSSLDIR)/misc/$$i; \
-	done
-	$(RM) $(DESTDIR)$(OPENSSLDIR)/openssl.cnf
+	done;
+	@set -e; for i in $(CONFS); \
+	do  \
+		echo $(RM) $(DESTDIR)$(OPENSSLDIR)/$$i; \
+		$(RM) $(DESTDIR)$(OPENSSLDIR)/$$i; \
+	done;
```
Comment (benlaurie) [inline apps/Makefile.in]: ditto

File: apps/apps.c [hunk 5]
```diff
@@ -235,6 +235,19 @@ int ctx_set_verify_locations(SSL_CTX *ctx, const char *CAfile,
     return SSL_CTX_load_verify_locations(ctx, CAfile, CApath);
 }
 
+int ctx_set_ctlog_list_file(SSL_CTX *ctx, const char *path)
+{
+    if (path == NULL) {
+        if (SSL_CTX_set_default_ctlog_list_file(ctx) <= 0) {
+            BIO_puts(bio_err, "Failed to load default Certificate Transparency "
```
Comment (benlaurie) [inline apps/apps.c]: Say what failed to load.

File: apps/Makefile.in (hunk 3 above)
Comment (RJPercival) [inline apps/Makefile.in]: Functionally, yes, but if anyone wants to add extra lines below this in the future, they don't have to notice that they need to add that semi-colon. I can take it out if you prefer though.

Comment (levitte) [inline apps/Makefile.in]: Hmmm?  The semi in itself doesn't matter, you can still add lines below.  The semi is only needed if you want to add to the same shell expression, and in that case, you'll likely need a backslash as well.

I'd remove those redundant semis, they're likely to be confusing more than anything else.

Comment (RJPercival) [inline apps/Makefile.in]: Ah they're all redundant? In that case, I'll remove them all (that is, the ones following "done").

File: apps/apps.c (hunk 5 above)
Comment (RJPercival) [inline apps/apps.c]: Unfortunately, the default file path is in crypto/include/internal/cryptlib.h. Am I ok reaching into a crypto internal header from apps? The comparable code for loading the default root certificates doesn't print an error message on failure, which is also an option.

File: apps/Makefile.in (hunk 3 above)
Comment (levitte) [inline apps/Makefile.in]: Those at the ends of lines that are not followed by a continuation backslash are redundant, yes.

Comment (RJPercival) [inline apps/Makefile.in]: Fixed.

File: apps/Makefile.in [hunk 6]
```diff
@@ -93,22 +94,30 @@ install:
...
-	@cp openssl.cnf $(DESTDIR)$(OPENSSLDIR)/openssl.cnf.new; \
-	chmod 644 $(DESTDIR)$(OPENSSLDIR)/openssl.cnf.new; \
-	mv -f  $(DESTDIR)$(OPENSSLDIR)/openssl.cnf.new $(DESTDIR)$(OPENSSLDIR)/openssl.cnf
+	@set -e; for i in $(CONFS); \
+	do \
+	(echo installing $$i; \
+	 cp $$i $(DESTDIR)$(OPENSSLDIR)/$$i.new; \
+	 chmod 644 $(DESTDIR)$(OPENSSLDIR)/$$i.new; \
+	 mv -f $(DESTDIR)$(OPENSSLDIR)/$$i.new $(DESTDIR)$(OPENSSLDIR)/$$i ); \
+	 done
```
Comment (levitte) [inline apps/Makefile.in]: Nit pick: take away that extra space before `done`.  `done` should be aligned with `do` 5 lines up.

And yeah, I know, it's indented that way at line 89.  Feel free to correct that as well ;-)

File: apps/apps.c (hunk 5 above)
Comment (benlaurie) [inline apps/apps.c]: Hmm. I guess that is a problem. OK, forget it.

**Comment by benlaurie** [comment]
+1

**Comment by richsalz** [comment]
DONE!

**Pull request closed without merging**